packages/rpc/src/WsClient.ts

```diff
@@ -62,87 +63,100 @@ export default class WsClient extends Client {
   private handleMessage = (data: MessageEvent<string>) => {
     const parsedData = JSON.parse(data.data) as unknown;
 
-    const response = rpcResponse.safeParse(parsedData);
+    const responses = z.array(rpcResponse).safeParse(parsedData);
 
-    if (!response.success) return;
+    if (!responses.success) return;
 
-    const { id } = response.data;
-
-    this.requestMap.get(id)?.resolve(response.data);
+    for (const response of responses.data) {
+      const { id } = response;
+      this.inFlightRequestMap.get(id)?.resolve(response);
+    }
   };
 
   private async connect(): Promise<WebSocket> {
-    const socket = new this.WebSocket(this.url);
+    this.shouldConnect = true;
+    this.emitter.dispatchEvent(new Event(CONNECTING));
+    const socket = new WebSocket(this.url);
     this.ws = socket;
 
     this.ws.addEventListener('message', this.handleMessage);
 
+    let connected = false;
+
+    const handleConnectionError = () => {
+      if (!connected) {
+        void this.handleDisconnect();
+      }
+    };
+
     // this event is also emitted if a socket fails to open, so all reconnection
     // logic will be funnelled through here
     this.ws.addEventListener('close', this.handleDisconnect, { once: true });
+    this.ws.addEventListener('error', handleConnectionError, { once: true });
+
+    await once(this.ws, 'open', { timeout: this.timeout });
 
-    await once(this.ws, 'open', { timeout: 30_000 });
+    connected = true;
 
     this.ws.addEventListener('error', () => {
       socket.close();
     });
 
     this.emitter.dispatchEvent(new Event(READY));
+    this.ws.removeEventListener('error', handleConnectionError);
 
     this.reconnectAttempts = 0;
 
     return socket;
   }
 
-  protected async send<const T extends RpcMethod>(
-    requests: JsonRpcRequest<T>[],
-  ): Promise<Response[]> {
-    const responses: Response[] = [];
-    for (const data of requests) {
-      const MAX_RETRIES = 5;
-      for (let i = 0; i < MAX_RETRIES; i += 1) {
-        let socket;
-        try {
-          socket = await this.connectionReady();
-        } catch {
-          // retry
-          continue;
-        }
-
-        socket.send(JSON.stringify(data));
-
-        const request = deferredPromise<unknown>();
-
-        this.requestMap.set(data.id, request);
-
-        const controller = new AbortController();
-        const result = await Promise.race([
-          sleep(30_000, { signal: controller.signal }).then(
-            () => ({ success: false, retry: false, error: new Error('timeout') }) as const,
+  protected async send(
+    requests: JsonRpcRequest<RpcMethod>[],
+    requestMap: RequestMap,
+  ): Promise<void> {
+    const MAX_RETRIES = 5;
+
+    let socket;
+    for (let i = 0; i < MAX_RETRIES; i += 1) {
+      try {
+        socket = await this.connectionReady();
+      } catch {
+        // retry
+        continue;
```

Comment:
I think it might make sense to sleep here for a short time?

Reply:
That happens up on line 55